vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#pragma once
+#include "vw/core/example.h"
+
+#include <list>
+#include <vector>
+#include <unordered_map>
+
+namespace VW
+{
+namespace reductions
+{
+VW::LEARNER::base_learner* eigen_memory_tree_setup(VW::setup_base_i& stack_builder);
+
+namespace eigen_memory_tree
+{
+
+using emt_feats = std::vector<std::pair<int, float>>;
+
+enum class emt_scorer_type
+{
+  random = 1,
+  distance = 2,
+  self_consistent_rank = 3,
+  not_self_consistent_rank = 4
+};
+
+enum class emt_router_type
+{
+  random = 1,
+  eigen = 2
+};
+
+float emt_median(std::vector<float>&);
+
+float emt_inner(emt_feats, emt_feats);
+
+emt_feats emt_scale(emt_feats, float);
+
+float emt_norm(emt_feats);
+
+emt_feats emt_normalize(emt_feats);
+
+emt_feats emt_scale_add(float const, emt_feats const, float const, emt_feats const);
```

Comment:
Looks like we are doing copying of emt_feats objects in all functions where emt_feats is part of signature.
Can it be changed to references or pointers?

Reply:
Yeah, that's a really good idea. Almost certainly. I'll make a performance pass over the code and improve this.

Reply:
I just finished an optimization pass to tune all function signatures (along with a few other optimizations). 

On a large test data set EMT now runs in about 4 seconds where it used to run in about 6 seconds.